exercises/palindrome-products/src/lib.rs

```diff
@@ -0,0 +1,12 @@
+pub type Palindrome = u64;
+pub fn get_palindrome_products(min: u64, max: u64) -> Vec<Palindrome>{
+    unimplemented!("Find all palindromic numbers which are products of numbers in the inclusive range ({}..{})", min, max)
+}
+
+pub fn min(palindromes: Vec<Palindrome>)->Option<Palindrome>{
```

Comment:
Accepting `Vec<Palindrome>` here transfers ownership of that vector, negating the potential performance improvement of computing once and then getting min/max.

It would be possible to accept a `&Vec<Palindrome>` instead and get that back, but that's not considered idiomatic: it locks the caller into a `Vec` implementation, when it may be more efficient for them to generate a slice in other ways.

It's therefore best to just accept a slice reference directly: `&[Palindrome]`.

This applies to `max` as well.

Reply:
I'm struggling a bit with the lifetimes. So when I return a slice over a vector i.e. palindrome.as_slice()  in get_palindrome_product() then I need to specify that the slice lives long enough. But I have no reference lifetime in my arguments. Should I give min and max as references or is there a more elegant approach?

Reply:
You don't really need to use lifetimes here for this to work. I've thrown up a quick branch to demonstrate: [lib.rs](https://github.com/coriolinus/exercism_rust/blob/palindrome-products/exercises/palindrome-products/src/lib.rs), [palindrome-products.rs](https://github.com/coriolinus/exercism_rust/blob/palindrome-products/exercises/palindrome-products/tests/palindrome-products.rs)

Reply:
There are a few performance improvements in there:

- [5](https://github.com/coriolinus/exercism_rust/blob/palindrome-products/exercises/palindrome-products/src/lib.rs#L5): `j` begins its lower bound at `i` instead of `min`. This eliminates roughly half the search space.
- [15, 19](https://github.com/coriolinus/exercism_rust/blob/palindrome-products/exercises/palindrome-products/src/lib.rs#L15): clone only the single item we care about, not the entire list. Doesn't matter so much for `Copy` types like `u64`, but for largeish things like strings, this really does speed things up.
- [26](https://github.com/coriolinus/exercism_rust/blob/palindrome-products/exercises/palindrome-products/src/lib.rs#L26): Perform `to_string()` only once. Do not copy the string. Producing a string requires heap allocation, which isn't quick. 
- [26](https://github.com/coriolinus/exercism_rust/blob/palindrome-products/exercises/palindrome-products/src/lib.rs#L26): Iterate over the bytes of the string instead of the characters. This is faster because when you use the `chars()` string iterator, it must produce a four-byte `char` on each iteration. For strings for which all chars are bytes, converting it into a byte vec and then iterating is faster. Naturally, if you can't guarantee that all chars in your string are in fact bytes, then you don't have any choice but to pay that price. Here, however, we know that all digits produced by the base-10 representation of a number are in fact a single byte wide.

Another performance improvement which I haven't implemented in this branch, but which should roughly halve the time required:

- Restructure tests to persist the list of palindrome products for min and max. That is, instead of having a pair of tests:

    ```rust
    #[test]
    fn smallest_palindrome_double_digits() {
        assert_eq!(min(&get_palindrome_products(10, 99)), Some(121));
    }
    #[test]
    fn largest_palindrome_double_digits() {
        assert_eq!(max(&get_palindrome_products(10, 99)), Some(9009));
    }
    ```

    We could persist the list of palindromes in a single test:

    ```rust
    #[test]
    fn palindromes_double_digits() {
        let pp = get_palindrome_products(10, 99);
        assert_eq!(min(&pp), Some(121));
        assert_eq!(max(&pp), Some(9009));
    }
    ```

    The part of these tests which takes the longest is searching a large number-space for palindrome products. By persisting the results of that search between the min and max checks, we can significantly reduce the time requires.

    There's a tradeoff to that approach, of course: it reduces the granularity of the tests. For long-running tests such as these, though, I believe that it's probably worth it.

Note that only the performance improvement which I didn't actually implement requires the use of a reference in the function API.

Reply:
Another unimplemented performance improvement: zip up only the first half of the byes checked in `is_palindrome` with the second half of the bytes checked there. Right now, each byte pair is checked twice; that's not necessary.

Reply:
It's worth mentioning that non-reference function arguments have fundamentally different semantics than in, say, go. In go, a non-reference function argument produces a copy, which would be expensive for a vector. In Rust, it _moves_ the variable. This is fast, but it means that the calling code no longer has access to that variable.